File: terminator/examples/cursor_example.rs
use serde_json;
use terminator::{AutomationError, Desktop, Selector};
use tracing::{error, info, warn};
use tracing_subscriber::EnvFilter;

#[tokio::main]
async fn main() -> Result<(), AutomationError> {
    // Initialize tracing
    tracing_subscriber::fmt()
        .with_env_filter(
            EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info")),
        )
        .init();

    println!("Starting script to find the 'Cursor' application and a button inside it.");

    // Create desktop automation instance
    let desktop = Desktop::new(false, false)?;

    // --- List all applications ---
    println!("\n--- Listing Running Applications ---");
    match desktop.applications() {
        Ok(apps) => {
            let app_names: Vec<String> = apps.iter().filter_map(|app| app.name()).collect();
            println!("Currently running applications: {:?}", app_names);
            info!("Found {} applications", apps.len());
        }
        Err(e) => {
            error!("Failed to get applications: {}", e);
            return Err(e);
        }
    }

    // --- Find the "Cursor" application ---
    println!("\n--- Finding 'Cursor' Application ---");
    match desktop.application("Cursor") {
        Ok(cursor_app) => {
            println!("Found Cursor application object:");
            let attrs = cursor_app.attributes();
            println!("  Name: {:?}", attrs.name);
            println!("  Role: {}", attrs.role);
            if let Ok(pid) = cursor_app.process_id() {
                println!("  Process ID: {}", pid);
            }

            // --- Find a button within "Cursor" using locator('role:button').first() ---
            println!("\n--- Finding a Button in 'Cursor' ---");

            // Create a role-based selector for buttons
            let button_selector = Selector::Role {
                // role: "button".to_string(),
                role: "AXButton".to_string(),
                name: None,
            };
            let locator = cursor_app.locator(button_selector)?;
            println!("Created locator for button with role 'button'");

            match locator
                .first(Some(std::time::Duration::from_secs(10)))
                .await
            {
                Ok(button) => {
                    println!("Found button object:");
                    let button_attrs = button.attributes();
                    println!("  Button Name: {:?}", button_attrs.name);
                    println!("  Button Role: {}", button_attrs.role);
                    if let Some(value) = &button_attrs.value {
                        println!("  Button Value: {}", value);
                    }
                    if let Some(description) = &button_attrs.description {
                        println!("  Button Description: {}", description);
                    }
                }
                Err(e) => match e {
                    AutomationError::ElementNotFound(_) => {
                        println!("No button found in Cursor application.");
                    }
                    _ => {
                        warn!("Error finding button: {}", e);
                        return Err(e);
                    }
                },
            }

            // --- Retrieve and print the entire window tree ---
            println!("\n--- Retrieving Window Tree for 'Cursor' ---");
            if let Ok(pid) = cursor_app.process_id() {
                match desktop.get_window_tree(pid, None, None) {
                    Ok(window_tree) => match serde_json::to_string_pretty(&window_tree) {
                        Ok(json) => {
                            println!("Window Tree JSON:\n{}", json);
                        }
                        Err(e) => {
                            warn!("Failed to serialize window tree: {}", e);
                        }
                    },
                    Err(e) => {
                        warn!("Failed to get window tree: {}", e);
                    }
                }
            } else {
                warn!("Could not determine process ID for Cursor application");
            }

            println!("\nScript finished successfully.");
        }
        Err(e) => {
            match e {
                AutomationError::ElementNotFound(_) => {
                    error!(
                        "Error: Could not find the requested element. Is 'Cursor' running? {}",
                        e
                    );
                }
                AutomationError::PlatformError(_) => {
                    error!(
                        "Error: A platform error occurred. Is 'Cursor' running? {}",
                        e
                    );
                }
                _ => {
                    error!("An unexpected error occurred: {}", e);
                }
            }
            return Err(e);
        }
    }

    Ok(())
}

File: bindings/nodejs/cursor_example.js
const {
    Desktop,
    ElementNotFoundError,
    PlatformError,
    Selector
} = require('.');
const fs = require('fs');

async function main() {
    console.log("Starting script to find the 'Cursor' application and a button inside it.");

    const desktop = new Desktop();

    try {
        // --- List all applications ---
        console.log("\n--- Listing Running Applications ---");
        const apps = desktop.applications();
        const appNames = apps.map(app => app.name());
        console.log("Currently running applications:", appNames);

        const appName = "Cursor";
        // const appName = "Google Chrome";
        // --- Find the application ---
        console.log(`\n--- Finding '${appName}' Application ---`);
        const app = desktop.application(appName);
        console.log("Found application object:");
        console.log(app);

        // --- Retrieve and print the entire window tree using getWindowTree ---
        console.log(`\n--- Listing windows for '${appName}' application ---`);
        const windows = await desktop.windowsForApplication(appName);
        console.log(`Found ${windows.length} window(s) for ${appName}.`);
        windows.forEach((win, idx) => {
            const attrs = win.attributes();
            const isMain = attrs.properties.AXMain === "true";
            console.log(`Window #${idx + 1}:`);
            console.log("  Name:", attrs.name || attrs.label);
            console.log("  Is Main:", isMain);
        });

        // Get the PID of the application
        const pid = app.processId();
        console.log(`${appName} PID: ${pid}`);
        
        // Optional: you can provide a window title filter (null means no filter)
        const selectedWindow = windows[0];
        const windowAttrs = selectedWindow.attributes();
        const windowId = selectedWindow.id();
        const windowTitle = windowAttrs.name || windowAttrs.label;
        console.log(`Window title: ${windowTitle} (id: ${windowId})`);

        const windowTree = desktop.getWindowTree(pid, windowTitle, null);

        const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
        const fileName = `data/window_tree_${pid}_${timestamp}.json`;
        fs.writeFileSync(fileName, JSON.stringify(windowTree, null, 2));
        console.log(`\nWindow tree written to ${fileName}`);
        
        // process.exit(0);

        // --- Find a button within "Cursor" using locator('role:button').first() ---
        console.log("\n--- Finding a Button in app ---");
        try {
            // You can tweak these locators to test various scenarios
            // const locator = app.locator('role:button');
            
            const locator1 = app.locator('button:Go Back (⌃-)');
            console.log("Locator 1:", locator1);
            const locator2 = app.locator(Selector.role('button', 'Go Back (⌃-)'));
            console.log("Locator 2:", locator2);

            const locator = locator2;

            // const locator = app.locator('button:Start Debugging (F5), use (⌥F1) for accessibility help');
            // const locator = app.locator('PopUpButton:Extensions');

            console.log("Locator for button:", locator);
            const button = await locator.first();
            console.log("Found button object:");
            console.log(button);

            // Find and log all buttons in the app
            // const allButtonsLocator = app.locator('role:button');
            const allButtonsLocator = app.locator(Selector.role('button'));
            const buttons = await allButtonsLocator.all();
            console.log(`Found ${buttons.length} button(s) in the Cursor application:`);
            buttons.forEach((button, idx) => {
                const attrs = button.attributes();
                console.log(`Button #${idx + 1}:`);
                console.log("  Name:", attrs.name);
                console.log("  Role:", attrs.role);
                if (attrs.value !== undefined) {
                    console.log("  Value:", attrs.value);
                }
                if (attrs.description !== undefined) {
                    console.log("  Description:", attrs.description);
                }
            });

        } catch (e) {
            if (e instanceof ElementNotFoundError) {
                console.log("No button found in Cursor application.");
            } else {
                throw e; // Rethrow for outer catch to handle
            }
        }

        console.log("\nScript finished successfully.");

    } catch (e) {
        if (e instanceof ElementNotFoundError) {
            // This could be because no button was found in Cursor.
            console.error("\nError: Could not find the requested element.", e.message);
        } else if (e instanceof PlatformError) {
            // This is likely because the "Cursor" application isn't running.
            console.error("\nError: A platform error occurred. Is 'Cursor' running?", e.message);
        } else {
            console.error('\nAn unexpected error occurred:', e);
        }
    }
}

main().catch(err => {
    console.error('\nFatal error:', err);
}); 